frontend/src/app/main/data/workspace/libraries.cljs

```diff
@@ -577,46 +578,84 @@
+               (when (seq rchanges)
+                 (dwc/commit-changes rchanges uchanges
+                                     {:commit-local? true
+                                      :file-id file-id})))))))
 
 (declare sync-file-2nd-stage)
 
 (defn sync-file
-  "Syhchronize the library file with the given id, with the current file.
-  Walk through all shapes in all pages that use some color, typography or
-  component of the library file, and copy the new values to the shapes.
-  Do it also for shapes inside components of the local file library."
-  [file-id]
+  "Syhchronize the given file from ghe given library. Walk through all shapes
+  in all pages in the file that use some color, typography or component of the
+  library, and copy the new values to the shapes. Do it also for shapes inside
+  components of the local file library."
+  [file-id library-id]
   (us/assert ::us/uuid file-id)
+  (us/assert ::us/uuid library-id)
   (ptk/reify ::sync-file
     ptk/UpdateEvent
     (update [_ state]
-      (if (not= file-id (:current-file-id state))
-        (assoc-in state [:workspace-libraries file-id :synced-at] (dt/now))
+      (if (not= library-id (:current-file-id state))
+        (assoc-in state [:workspace-libraries library-id :synced-at] (dt/now))
         state))
 
     ptk/WatchEvent
     (watch [_ state stream]
-      (log/info :msg "SYNC-FILE" :file (if (= file-id (:current-file-id state)) "local" (str file-id)))
-      (let [local-library (dwlh/get-local-library state)
-            library-changes [(dwlh/generate-sync-library :components file-id state)
-                             (dwlh/generate-sync-library :colors file-id state)
-                             (dwlh/generate-sync-library :typographies file-id state)]
-            file-changes    [(dwlh/generate-sync-file :components file-id state)
-                             (dwlh/generate-sync-file :colors file-id state)
-                             (dwlh/generate-sync-file :typographies file-id state)]
+      (log/info :msg "SYNC-FILE"
+                :file (dwlh/pretty-file file-id state)
+                :library (dwlh/pretty-file library-id state))
+      (let [file            (dwlh/get-file state file-id)
+            library-changes [(dwlh/generate-sync-library file-id :components library-id state)
+                             (dwlh/generate-sync-library file-id :colors library-id state)
+                             (dwlh/generate-sync-library file-id :typographies library-id state)]
+            file-changes    [(dwlh/generate-sync-file file-id :components library-id state)
+                             (dwlh/generate-sync-file file-id :colors library-id state)
+                             (dwlh/generate-sync-file file-id :typographies library-id state)]
             rchanges (d/concat []
                                (->> library-changes (remove nil?) (map first) (flatten))
                                (->> file-changes (remove nil?) (map first) (flatten)))
```

Comment:
The same here: the repeated logic can be expressed using transducers:

```clojure
      xf-fcat  (comp (remove nil?) (map first) (mapcat identity))
      rchanges (d/concat []
                         (sequence xf-fcat library-changes)
                         (sequence xf-fcat file-changes))
```
NOTE: flatten is replaced with `mapcat identity` assiming you have only a single level of flattering

Reply:
done